Add smartapi kg

sengineer0 opened the pull request:
In order to provide an api for query MetaKG with scalable feature, this PR try to implement these things:
- [x] Copy smartapi_kg module from [bte-py repo](https://github.com/newgene/bte-py/tree/main/biothings_explorer/smartapi_kg)
- [x] Add MetaKG model for storing meta_kg data in ElasticSearch
- [x] Add MetaKg controller for providing some neccessary functions, like: get_all, get, find, delete, save, .... meta_kg documents
- [x] Refactor model, controller modules to deduplicate code
- [x] Add refresh_metakg admin function
- [x] Add an api for querying MetaKG document
- [ ] Add tests for MetaKG model, MetaKG controller, refresh_metakg

 
Ref: https://github.com/SmartAPI/smartAPI/issues/179

sengineer0 committed:
Add smartapi_kg module for harvest metakg

sengineer0 committed:
Add model document for MetaKg; refactor models module

sengineer0 committed:
Add controller for MetaKG; refactor controller module

sengineer0 committed:
Add refresh_metakg admin function

sengineer0 committed:
Add api for query metakg

sengineer0 committed:
Update incorrect references; Update tests

sengineer0 committed:
Add flake8 config; update gitignore

newgene commented on src/controller/smartapi.py:
```diff
@@ -0,0 +1,182 @@
+            raise ValueError("Slug contains invalid characters.")
+
+
+class SmartAPIEntity(AbstractWebEntity, Mapping):
```
Why is it renamed from `SmartAPI` to `SmartAPIEntity`? Seems not necessary to me @sengineer0

sengineer0 replied:
After I added a new model to store MetaKG, we have controllers, models for MetaKG, SmartAPI. So I think it will make the code more readable and we can easily determine what obj type is.
@newgene

newgene commented on src/model/smaprtapi.py:
```diff
@@ -0,0 +1,53 @@
+"""
```
This file name has a typo: `smaprtapi.py` @sengineer0

sengineer0 replied:
oops, my bad, I will fix it.

newgene reviewed (commented):
@sengineer0 let's not copy the original `smartapi_kg` here. We should simplify most of the original code and probably put it as a `utils.metakg` submodule.

Let's first go over how this `utils.metakg` should work here.

sengineer0 commented:
> @sengineer0 let's not copy the original `smartapi_kg` here. We should simplify most of the original code and probably put it as a `utils.metakg` submodule.
> 
> Let's first go over how this `utils.metakg` should work here.

So, basically you only want the minimal code should be copied, right?
@newgene

newgene commented:
yes, keep only what's needed for smartapi application. You can also rewrite to make it more like a typical Python class, v.s. the current structure mapped from its nodejs version. @sengineer0

sengineer0 commented:
> yes, keep only what's needed for smartapi application. You can also rewrite to make it more like a typical Python class, v.s. the current structure mapped from its nodejs version. @sengineer0

ok, I will update it soon.
@newgene

sengineer0 committed:
Refactor smartapi_kg module
- Reduce smartapi_kg module to only essential files
- Rename smartapi_kg to metakg
- Move metakg module to utils module

sengineer0 commented:
@newgene 
I did a refactor smartapi_kg to remove any unneccessary logic, rename it to metakg and move it to under utils module
Please have a check

newgene commented on src/utils/metakg/config.py:
```diff
@@ -0,0 +1 @@
+SMARTAPI_URL = "https://smart-api.info/api/query?q=tags.name:translator&size=500&fields=paths,servers,tags,components.x-bte*,info,_meta&meta=1"
```
should just use SmartAPI model to get the API metadata directly, not from this URL, since we are now in the SmartAPI application itself.

sengineer0 replied:
oh ok, I just don't know they are the same. I will update my PR.

newgene commented on src/utils/metakg/loader.py:
```diff
@@ -0,0 +1,29 @@
+        if url:
+            self._url = url
+
+    def fetch(self):
```
since we will not use SMARTAPI_URL, this part should be rewritten completely

newgene commented on src/utils/metakg/parser/index.py:
```diff
@@ -0,0 +1,80 @@
+from .endpoint import Endpoint
```
the name `index.py` is from the node version, does not make sense in Python. I would put it into `__init__.py`

newgene reviewed (commented):
* Don't think we need `data/predicates.json` and the whole `data` folder in the repo

sengineer0 commented:
> should just use SmartAPI model to get the API metadata directly, not from this URL, since we are now in the SmartAPI application itself.

@newgene As you want to get metadata directly from SmartAPI model, I suggest that we should bring the logic extract metadata to controller, model instead of using utilsmeta. It can help us reduce the complexity of the logic.
What do you think?

newgene commented:
@sengineer0 that's possible, depending on how complex the logic is. If it becomes too complex, it's better to separate it out.

One question though, in the current `metakg` module, does it still support `.construct_MetaKG(includeReasoner=True)`? For each of those reasoner API entries, we actually need to query its `/meta_knowledgegraph` endpoint to retrieve the `metakg` of this API.

sengineer0 commented:
> @sengineer0 that's possible, depending on how complex the logic is. If it becomes too complex, it's better to separate it out.
> 
> One question though, in the current `metakg` module, does it still support `.construct_MetaKG(includeReasoner=True)`? For each of those reasoner API entries, we actually need to query its `/meta_knowledgegraph` endpoint to retrieve the `metakg` of this API.

I already brang the logic to controller and model, but I deleted include_reasoner as I didn't know you want it.
So, today I will add it back :)

@newgene

sengineer0 committed:
Extract metadata directly from SmartAPIDoc instead of fetching api; refactor MetaKG util module;

sengineer0 committed:
add include_reasoner for refresh_metakg

newgene commented on src/admin.py:
```diff
@@ -166,11 +170,21 @@ def debug_uptime(_id, endpoint=None):
+    docs = meta_kg.get_associations()
+    for doc in docs:
+        meta_kg_doc = MetaKGDoc(**doc)
+        meta_kg_doc._raw = decoder.compress(json.dumps(doc).encode())
```
@sengineer0 I actually don't think we need to store a `_raw` field for metakg association docs. In `SmartAPIEntity` case, we did some transformation (for better ES indexing) so we want to store a blob of the raw doc in `_raw` field to get the exact raw data.

sengineer0 replied:
oh, ok. Let's me update to remove `_raw` field in MetaKGDoc
@newgene

sengineer0 replied:
It's updated. @newgene

newgene commented on src/model/base.py:
```diff
@@ -35,44 +25,16 @@ class StatMeta(InnerDoc):
-
-    _meta = Object(UserMeta, required=True)
+class BaseDoc(Document):
     _status = Object(StatMeta)
     _raw = Binary()
```
@sengineer0 I don't think metakg model needs either _status or _raw field.

sengineer0 replied:
@newgene Sure, will remove them.

sengineer0 replied:
It's updated. @newgene

newgene reviewed (commented):
Also want to mention that MetaKGDoc will only be updated through our daily update script, won't be updated directly from the user. This is different from SmartAPIDoc, which is from user's submission.

newgene commented on src/config.py:
```diff
@@ -77,6 +77,7 @@
     (r'/api/build/?', 'biothings.web.handlers.MetadataSourceHandler'),
     (r'/api/status/?', 'biothings.web.handlers.StatusHandler'),
     (r'/api/suggestion/?', 'handlers.api.ValueSuggestionHandler'),
+    (r'/api/metakg/?', 'handlers.api.MetaKGHandler'),
```
@sengineer0 we should also have a `/api/metakg/query` endpoint, just like `/api/query` is for querying SmartAPIDoc index.

sengineer0 replied:
ok let's me add it

sengineer0 replied:
@newgene It's updated.

sengineer0 commented:
> Also want to mention that MetaKGDoc will only be updated through our daily update script, won't be updated directly from the user. This is different from SmartAPIDoc, which is from user's submission.

@newgene  Sure, I only added an admin function, and can be used only by our code, or by importing manually.

sengineer0 committed:
rename metakg.index to metakg.api

sengineer0 committed:
remove unneccessary fields on MetaKGDoc; move some methods which only used by SmartAPIEntity to SmartAPIEntity

sengineer0 committed:
Use biothings.web.handlers.QueryHandler for querying MetaKG

newgene commented on src/model/metakg.py:
```diff
@@ -0,0 +1,44 @@
+class SmartAPIInnerDoc(InnerDoc):
+    metadata = Keyword(required=True)
+    id = Keyword(required=True)
+    ui = Keyword()
```
`metadata` and `ui` are the URL field, we don't need to index them

sengineer0 replied:
it's updated @newgene

newgene commented on src/model/metakg.py:
```diff
@@ -0,0 +1,44 @@
+class APIInnerDoc(InnerDoc):
+    name = Text()
+    smartapi = Object(SmartAPIInnerDoc)
+    x_translator = Object(name="x-translator")
```
we should index the nested fields under `x-translator`

newgene replied:
you can give me an example, so I can tell you exactly which to index, which is not. In general, most of nested fields should be indexed.

sengineer0 replied:
This is an example for non-trapi record

```
{
  "query_operation": {
    "params": {
      "variantid": "{{ queryInputs }}%23%23"
    },
    "path_params": [
      "variantid"
    ],
    "path": "/entity/litvar/{variantid}",
    "method": "get",
    "server": "https://www.ncbi.nlm.nih.gov/research/bionlp/litvar/api/v1",
    "tags": [
      "variant",
      "translator"
    ]
  },
  "association": {
    "input_id": "DBSNP",
    "input_type": "SequenceVariant",
    "output_id": "NCBIGene",
    "output_type": "Gene",
    "predicate": "is_sequence_variant_of",
    "source": "infores:dbsnp",
    "api_name": "LitVar API",
    "smartapi": {
      "id": "dca415f2d792976af9d642b7e73f7a41",
      "meta": {
        "date_created": "2020-12-07T23:08:24.199007+00:00",
        "last_updated": "2023-01-31T08:03:44.557076+00:00",
        "url": "https://raw.githubusercontent.com/NCATS-Tangerine/translator-api-registry/master/litvar/smartapi.yaml",
        "username": "newgene"
      }
    },
    "x-translator": {
      "component": "KP",
      "team": [
        "Service Provider"
      ]
    }
  },
  "response_mapping": {
    "is_sequence_variant_of": {
      "NCBIGene": "gene.id",
      "biolink:source_web_page": "links.url"
    }
  },
  "tags": [
    "variant",
    "translator"
  ]
}
```

sengineer0 replied:
and this one for trapi

```
{
    "association": {
        "input_type": "biolink:GeneOrGeneProduct",
        "output_type": "biolink:ChemicalEntity",
        "predicate": "biolink:has_part",
        "api_name": "Ontology-KP API",
        "smartapi": {
            "id": "2ca2daadd989387f1e8f7121fa214eda",
            "meta": {
                "date_created": "2022-02-28T20:50:09.725794+00:00",
                "last_updated": "2023-01-31T08:03:03.942189+00:00",
                "url": "https://ontology-kp.apps.renci.org/docs/docs.yaml",
                "username": "balhoff"
            }
        },
        "x-translator": {
            "component": "KP",
            "team": [
                "Standards Reference Implementation Team"
            ]
        }
    },
    "tags": [
        "translator",
        "trapi",
        "bte-trapi"
    ],
    "query_operation": {
        "path": "/query",
        "method": "post",
        "server": "https://ontology-kp.apps.renci.org",
        "path_params": null,
        "params": null,
        "request_body": null,
        "support_batch": true,
        "input_separator": ",",
        "tags": [
            "translator",
            "trapi",
            "bte-trapi"
        ]
    }
}
```

@newgene

newgene replied:
you may consider using the dynamic mapping approach as in SmartAPI model:

https://github.com/newgene/smartAPI/blob/master/src/utils/indices.py#L20

sengineer0 replied:
I am not really understand this approach, lets me check it first and I will ask you if there is any question.
@newgene

sengineer0 replied:
it's updated @newgene

newgene commented on src/model/metakg.py:
```diff
@@ -0,0 +1,44 @@
+
+
+class MetaKGDoc(BaseDoc):
+    subject = Text()
+    object = Text()
+    predicate = Text()
+    provided_by = Text()
```
All these text fields should also be indexed as keyword, similar to how we did here:

https://github.com/newgene/smartAPI/blob/master/src/utils/mapping.json#L54-L68

that means we will have `subject` indexed as Text for fulltext search, and `subject.raw` as Keyword for aggregation etc. Text field cannot be used for aggregation by default.

sengineer0 replied:
did you mean that these fields should be defined like this:
```
subject = Text(field={'raw': Keyword()})
```

@newgene

newgene replied:
👍looks good to me.

newgene commented on src/controller/smartapi.py:
```diff
@@ -0,0 +1,287 @@
+        BULK = 10000
+        count_docs = cls.count()
+        for i in range(math.ceil(count_docs / BULK)):
+            entities = cls.get_all(BULK, i * BULK)
```
see this URL from the original smartapi_kg.js package:

https://github.com/biothings/BioThings_Explorer_TRAPI/blob/main/src/controllers/cron/update_local_smartapi.js#L210

We actually have a filter for SmartAPI entries for `tags.name` contains `translator`, so we should apply the same filter here using the SmartAPI model.

sengineer0 replied:
ok, I will add that filter soon

sengineer0 replied:
it's updated @newgene

newgene commented on src/model/metakg.py:
```diff
@@ -0,0 +1,44 @@
+from .base import BaseDoc
+
+
+ES_INDEX_NAME = 'metakg_docs'
```
let's call it `smartapi_metakg_docs` instead (so that searching "smartapi", both indices will come up)

sengineer0 replied:
ok. it will be updated soon

sengineer0 replied:
it's updated @newgene

newgene commented on src/controller/metakg.py:
```diff
@@ -0,0 +1,34 @@
+            metadatas = parser.get_non_TRAPI_metadatas(original_data, extra_data)
+            if include_reasoner:
+                metadatas += parser.get_TRAPI_metadatas(original_data)
+            for metadata in metadatas:
+                cls.create_doc(metadata)
```
possible to use bulk save here? Will it improve performance significantly?

This might be relevant: https://github.com/elastic/elasticsearch-dsl-py/issues/403

newgene replied:
another issue here, if we continue saving doc one at a time, is many line of elasticsearch logger output like this:

```
INFO:elasticsearch:POST http://localhost:9200/metakg_docs/_doc [status:201 request:0.049s]

```
we probably want to suppress that for cleaner log output.

sengineer0 replied:
> possible to use bulk save here? Will it improve performance significantly?

Let me check first, if it is possible, I will update to use bulk save.

> another issue here, if we continue saving doc one at a time, is many line of elasticsearch logger output like this:

Sure, there logs will be suppressed.

sengineer0 replied:
it's updated @newgene

newgene commented on src/controller/metakg.py:
```diff
@@ -0,0 +1,34 @@
+    MODEL_CLASS = MetaKGDoc
+
+    @classmethod
+    def create_by_smartapis(cls, smartapi_entities, include_reasoner=False):
+        for smartapi_entity in smartapi_entities:
```
since we are now in the SmartAPI application itself. I don't see any reason we want to process metadata in a batch here as `smartapi_entities`. We can just loop through docs from SmartAPI model, and handle one API at a time:

```
for smartapi_entity in smartapi_all:
     if tags.name contains translator:
          if TRAPI API:
                parser.get_TRAPI_metadatas
          elif x_bte API:
                parser.get_non_TRAPI_metadatas
          else:
                just skip
          bulk_save all metadata from one API
```

newgene replied:
for easier debugging and cleaner code structure, you might want to have a method of parsing out metakg metadata for a given API id, which covers the above `if...elif...else` block:

```
parser.get_metakg(smartapi_id)
```

sengineer0 replied:
> since we are now in the SmartAPI application itself. I don't see any reason we want to process metadata in a batch here as smartapi_entities. We can just loop through docs from SmartAPI model, and handle one API at a time:

Of course, I loaded and process metadata for all stored SmartAPI. but I just think doing in batch to prevent loading too many data into memory at a time, so I split them into small set, and process each one in a batch.

If you don't want to do that, I will remove the batch logic, and process all SmartAPI in one time.

@newgene

newgene replied:
@sengineer0 I actually think the opposite. Because we are within the SmartAPI application, so you can just process one SmartAPI entry at a time, no need to be in a batch. That way we can further reduce memory usage. Definitely not loading all entries in memory and processing them in one batch, which will make it worst :-)

However, I think for the controller method, you can keep ```create_by_smartapis``` but create another one ```create_by_smartapi``` for processing one at a time, even another method like `create_by_smartapi_id`. These are very useful for debugging purposes.

sengineer0 replied:
that makes seen. I will update as you recommended :)

@newgene

sengineer0 replied:
it's updated @newgene

newgene commented on src/admin.py:
```diff
@@ -166,11 +168,15 @@ def debug_uptime(_id, endpoint=None):
         smartapi.save()
 
 
+def refresh_metakg(include_reasoner=False):
```
@sengineer0 let's set `include_reasoner` as True by default.

sengineer0 replied:
Sure, will update soon.

sengineer0 replied:
it's updated @newgene

sengineer0 committed:
Some minor changes as reflecting comments
- remove unnecessary fields from models
- using the dynamic mapping approach for indexing nested fields for
  MetaKGDoc
- remove unnecessary batch logic
- change MetaKG index name
- Using bulk to create MetaKG docs
- suppress ElasticSearch info logs when run refresh_metakg
- make include_reasoner as True by default

newgene committed:
refactor metakg methods

newgene committed:
updated metakg index name

newgene committed:
refactor: :truck: revert SmartAPIEntity controller to SmartAPI

newgene committed:
feat: improve the logic of getting server url

newgene committed:
style: fourmat fixes

newgene committed:
feat: add more logs in refresh_all_metakg
